feat(funnels): add analytics events to track click / browse / search

LinHuiqing opened the pull request:
## Problem

Currently we are limited in our ability to track how users interact with our site. To make better informed design decisions, we then need to gather data on how users are using the site.

Closes #294. Refer to issue for more details.

## Solution

Add custom events to be fired for Google Analytics (GA) and FullStory (FS) to catch. FS is then configured to show the conversion funnels on the dashboard.

How events captured are combined to create conversion funnels:

![AskGov Conversion Funnels](https://user-images.githubusercontent.com/37061143/133715089-4149f143-8297-4dc9-b9f7-e5875949b84b.jpeg)

**Features**:

- Add 'Click Tag' events for both GA and FS. These events are tagged with the following details:
  - String of the tag which is clicked.
  - Time at which the tag is clicked starting from when app is loaded.
- Add 'Browse' events for both GA and FS.
  - Agency page on which browsing was done, identified by the agency's short name.
  - Percentage of the page which was scrolled.

**Improvements**:

- Refactor both `client/src/services/googleAnalytics.tsx` and `client/src/contexts/googleAnalytics.tsx` so that `googleAnalytics` can be used in TypeScript files.
- Refactor `TagBadge` component to TypeScript.

## Tests

With an IP address that is not blocked from recording by FS (refer to [this page](https://app.fullstory.com/ui/12P870/settings/recording)), navigate to the home page.

### Browse Events

1. Scroll a distance then navigate to another page (possibly by clicking into a question).
2. Close the page.
3. On FS, a new session should be registered. This session should include 'Browse' events which have the agency's short name for the agency page that is browsed and the percentage of the page which has been scrolled.
4. On GA, 'Browse' events should be captured on the Realtime dashboard.

### Click Tag Events

1. Click on tags, either those under questions or on the side panel.
2. Close the page.
3. On FS, a new session should be registered. This session should include 'Click Tag' events which have the names of the tags which are clicked on and the time after page load which the tag click was made.
4. On GA, 'Click Tag' events should be captured on the Realtime dashboard.

LinHuiqing committed:
feat(funnels): capture click tag events

- add click tag to user events
- send click tag user event to GA and FullStory when tags are
clicked
- refactor TagBadge component to Typescript

LinHuiqing committed:
feat(funnels): capture time of first tag click

- modify tag click event captured to include time before event (since app load)
- refactor
googleAnalytics modules to include isFirstSearch and appLoadTime in context props

re #294

LinHuiqing committed:
feat(funnels): add click tag event to tag panel

re #294

LinHuiqing committed:
feat(funnels): add post list scroll percentage event

- created 'Browse' event label and maxScroll tracker on GA context and service
- add onscroll
listener to update maximum scroll position on pages with PostList
- send event with maximum scroll
percentage when component is unmounted
- note: due to browser compatibility issues, scroll
percentage might differ for the same screen size and code is added to limit percentage to 100

re #294

mantariksh commented on client/src/components/PostList/PostList.component.jsx:
```diff
@@ -18,6 +19,54 @@ const PostList = ({ posts, defaultText, alertIfMoreThanDays, showViews }) => {
+  }
+
+  const sendMaxScrollToAnalytics = (maxScrollPossible) => {
+    maxScrollPossible = Math.max(maxScrollPossible, googleAnalytics.maxScroll) // this is required to restrict the percentage to 100
```
why would `googleAnalytics.maxScroll` ever exceed `maxScrollPossible`? wouldn't that just make `maxScrollPercentage` 1?

LinHuiqing replied:
If I'm not wrong, there are some browser compatibility issues which cause the total scroll height (i.e. `maxScrollPossible`) to differ on different browsers even with the same screen size (refer to [this post](https://stackoverflow.com/questions/17688595/finding-the-maximum-scroll-position-of-a-page)). There is a better API to track this ([window.scrollMaxY](https://developer.mozilla.org/en-US/docs/Web/API/Window/scrollMaxY)), but it's only available on Firefox so...

When the `maxScroll` exceeds `maxScrollPossible`, its not a lot usually within 20%. Since we're just using this as a gauge to see whether users browse, I think the precision doesn't have to be very high?

mantariksh replied:
I see! can we leave some documentation here (can link to the relevant SO post also) for future developers?

mantariksh commented on client/src/components/PostList/PostList.component.jsx:
```diff
@@ -18,6 +19,54 @@ const PostList = ({ posts, defaultText, alertIfMoreThanDays, showViews }) => {
+  useEffect(() => {
+    resetMaxScrollAndAddListener()
+    const pageHeight = Math.max(
+      document.body.scrollHeight,
+      document.body.offsetHeight,
+      document.documentElement.clientHeight,
+      document.documentElement.scrollHeight,
+      document.documentElement.offsetHeight,
```
what's the difference between all of these? is there no single source of truth for the height of the viewport?

LinHuiqing replied:
Sadly, there isn't :( Unless we assume that users only use certain browsers. Refer to [this post](https://stackoverflow.com/questions/17688595/finding-the-maximum-scroll-position-of-a-page).

mantariksh replied:
thanks! can link to the SO post in a comment in the code pls?

mantariksh commented on client/src/components/SearchBox/SearchBox.component.jsx:
```diff
@@ -69,7 +69,7 @@ const SearchBox = ({
       timeToFirstSearch_int: timeToFirstSearch,
     })
-    googleAnalytics.setIsFirstSearch(true)
+    googleAnalytics.isFirstSearch = true
```
given that `isFirstSearch` is state in the `GoogleAnalyticsProvider`, we should be using the `setIsFirstSearch` setter so that the change propagates correctly - see [useState docs](https://reactjs.org/docs/hooks-state.html)

LinHuiqing replied:
I changed it to a variable, because I'm not very convinced of the use of states when the value is only changed once and is only read after.

mantariksh commented on client/src/components/TagBadge/TagBadge.component.tsx:
```diff
@@ -0,0 +1,50 @@
+  const googleAnalytics = useGoogleAnalytics()
+
+  const sendClickTagEventToAnalytics = () => {
+    const timeToTagClick = new Date().getTime() - googleAnalytics.appLoadTime
```
no need to change but just FYI these are equivalent
```suggestion
    const timeToTagClick = Date.now() - googleAnalytics.appLoadTime
```

LinHuiqing replied:
Nice, looks cleaner, thanks! Will change it together with other places where this is done.

mantariksh commented on client/src/components/TagBadge/TagBadge.component.tsx:
```diff
@@ -0,0 +1,50 @@
+      timeToTagClick,
+    )
+    googleAnalytics.sendTiming(
+      'User',
```
noticed we are slotting this into the same category as `sendUserEvent` - would providing a different category make the GA console more organised? up to you, I'm not super familiar with this

LinHuiqing replied:
I think its fine because we can search by the event label on GA? Either way should be fine, but User is a pretty big category so I just assumed it belonged there.

mantariksh reviewed (commented):
for your consideration - what do you think about keeping the `maxScroll` state in `PostList.component` itself, perhaps via a [ref](https://reactjs.org/docs/hooks-reference.html#useref), instead of in the `googleAnalytics` service?

a couple of reasons for this:
1. single responsibility principle. imo the GA service should just be in charge of firing GA events, not keeping track of state in other components.
2. imo the GA service is already keeping track of too much state in other components, which makes things confusing. when reading the GA service in isolation, it is not at all obvious what `maxScroll` is doing. imagine if every single component which fired GA events started sharing state with the GA service - how would we know what each variable represents?

LinHuiqing commented:
> for your consideration - what do you think about keeping the `maxScroll` state in `PostList.component` itself, perhaps via a [ref](https://reactjs.org/docs/hooks-reference.html#useref), instead of in the `googleAnalytics` service?
> 
> a couple of reasons for this:
> 
> 1. single responsibility principle. imo the GA service should just be in charge of firing GA events, not keeping track of state in other components.
> 2. imo the GA service is already keeping track of too much state in other components, which makes things confusing. when reading the GA service in isolation, it is not at all obvious what `maxScroll` is doing. imagine if every single component which fired GA events started sharing state with the GA service - how would we know what each variable represents?

For this, I started with the approach which you suggested actually, but I changed to the approach I am using now because googleAnalytics was storing the appLoadTime (so I assume that was the preferred approach). Otherwise, the appLoadTime could have been a ref too I think?

Anyway, yea I agree its quite bloated and should be in the component itself as it does not need to be shared.

mantariksh commented:
>Otherwise, the appLoadTime could have been a ref too I think?

agreed, although I think `appLoadTime` is a bit more forgivable since it's a genuinely app-wide variable, whereas `maxScroll` is actually specific to `PostList`

LinHuiqing committed:
refactor(funnels): address comments made to PR #319

- add comments referring to SO post to justify scroll height related code
-  changed new
Date().getTime() to Date.now()
- remove maxScroll variable from `googleAnalytics` to `postList`
instead

re #294

LinHuiqing committed:
refactor(searchbox): remove hardcoded string for FullStory timing event

LoneRifle commented on client/src/components/PostList/PostList.component.jsx:
```diff
@@ -18,6 +19,55 @@ const PostList = ({ posts, defaultText, alertIfMoreThanDays, showViews }) => {
+    if (maxScroll.current !== 0) sendMaxScrollToAnalytics(maxScrollPossible)
+  }
+
+  useEffect(() => {
+    resetMaxScrollAndAddListener()
+    // Due to browser compatibility issues, there is no single source of truth for maximum scroll height:
+    // https://stackoverflow.com/questions/17688595/finding-the-maximum-scroll-position-of-a-page
+    const pageHeight = Math.max(
+      document.body.scrollHeight,
+      document.body.offsetHeight,
+      document.documentElement.clientHeight,
+      document.documentElement.scrollHeight,
+      document.documentElement.offsetHeight,
+    )
+    const maxScrollPossible = pageHeight - window.innerHeight
+    return () => removeListenerAndSendEvent(maxScrollPossible)
+  }, [])
```
Some commentary of what this does would be helpful. It's not immediately obvious that the hook installs a callback that gets triggered only when the component unloads

LoneRifle reviewed (approved):
lgtm otherwise, nicely done!

LinHuiqing committed:
docs(postlist): comment on implementation of browse event trigger
